### DISCUSSION TITLE
Schedule and Logic Behind Velero Expired Backup Deletion Frequency

### DISCUSSION THREAD

**Discussion by jbaskeen** [vmware-tanzu/velero#5913] [Community support Q&A] [answered]
I am looking to understand how backup deletion runs are implemented as I have not seen any documentation detailing the topic. Backup expiration is managed by the corresponding schedule `ttl`. This is clear. However, when a backup expires I see it sit for a while as expired before it gets cleaned up. Eventually an expired backup is deleted but I have not yet determined what the frequency of that "job" is and whether that event occurs on a regular interval. As a result, I am looking to know....

1. What is the logic implemented by Velero server for deleting an expired backup?
2. What is the frequency of backup deletions as managed by Velero server?
3. Is this frequency controllable by the user/admin? Is it possible to kick off the expired backup deletion manually, without manually deleting the backup myself?

Thanks for the support!

**Comment by sseago** [comment, marked as answer]
The GC (garbage collection) controller manages this. It runs periodically and creates Backup deletion requests for each expired backup it finds. By default it runs once per hour, but can be configured using the `garbage-collection-frequency` server parameter.

**Reply by jbaskeen** [reply]
Thanks @sseago, this is exactly what I need, for the garbage collection to run more frequently in some cases. Other than the source code, is there someplace where the server flags and their descriptions etc are documented?

**Reply by sseago** [reply]
The CLI itself documents the actual parameters. For server params, `velero help server` lists the accepted flags. Among other things, it includes the following in its output:
```
      --garbage-collection-frequency duration               How often garbage collection is run for expired backups.
```

**Reply by jbaskeen** [reply]
Ever considered adding some level of info to the docs about configurable options? I usually find myself digging through service documentation when evaluating a solution before even downloading a cli and inspecting its flags etc. But I do understand why some teams don't take this approach and why others do. 

Another area where I struggle with Velero docs is the Plugins section, for example. I have a potential use case for a plugin but there's not enough info in the documentation to tell me how integration works in detail and the workflow with a plugin. And the examples seem very basic. As a result, I am less likely to develop the plugin since I don't have all the relevant details, even though I want to. But that's a separate scope I suppose.

**Reply by sseago** [reply]
@jbaskeen So there is an open issue about this -- https://github.com/vmware-tanzu/velero/issues/4905
It looks like it hasn't been considered a priority, and there's some concern that if added, it might quickly fall out-of-date unless something automated is put in place, but feel free to comment there to express interest. One reason this was never pursued was that it wasn't clear that we had any interest among users in having this, which may not be the case anymore.

**Reply by jbaskeen** [reply]
Awesome, thanks for the info. Will check it out, cheers

**Comment by sseago** [comment]
You can kick off an individual backup deletion manually by running `velero backup delete`, but I don't think there's a way to trigger immediate reconcile of the garbage collection controller, but you can configure it to run more often, as mentioned above.
